File: 543-diameter-of-binary-tree/diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int diameterOfBinaryTreeHelper(TreeNode* root, int &diameter){
        if(!root) return 0;
        int lft= diameterOfBinaryTreeHelper(root->left, diameter);
        int rgt= diameterOfBinaryTreeHelper(root->right, diameter);
        diameter= max(diameter, lft+rgt);
        return 1+max(lft, rgt);
    }
    int diameterOfBinaryTree(TreeNode* root) {
        int diameter=0;
        diameterOfBinaryTreeHelper(root, diameter);
        return diameter;
    }
};

File: 543-diameter-of-binary-tree/diameter-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int diameter;
    int height(TreeNode root) {
        if(root==null) return -1;
        int left= 1+height(root.left);
        int right= 1+height(root.right);
        diameter= Math.max(left+right, diameter);
        return Math.max(left,right);
    }
    public int diameterOfBinaryTree(TreeNode root) {
        if(root==null) return 0;
        diameter=0;
        height(root);
        return diameter;
    }
}